### ISSUE TITLE
Learning rate

### ISSUE THREAD

**Issue by Jean-Baptiste-Camps** [hipster-philology/pandora#10]
I have noticed something weird while retraining a model, and I think it has to do with learning rate. During the first training on the corpus (halve lr at 50), after 100 epochs, my results were seemingly hitting a ceiling: 

    -> epoch  100 ...
    - Lowering learning rate > was: 0.004999999888241291 , now: 0.0025
    Epoch 1/1
    37853/37853 [==============================] - 200s - loss: 2.6546 - lemma_out_loss: 1.7012 - pos_out_loss: 0.9534   
    ::: Train scores (lemmas) :::
    +	all acc: 0.8658758883047579
    +	kno acc: 0.8658758883047579
    +	unk acc: 0.0
    ::: Dev scores (lemmas) :::
    +	all acc: 0.8330515638207946
    +	kno acc: 0.8961684011352885
    +	unk acc: 0.30357142857142855
    ::: Train scores (pos) :::
    +	all acc: 0.8463529971204395
    +	kno acc: 0.8463529971204395
    +	unk acc: 0.0
    ::: Dev scores (pos) :::
    +	all acc: 0.8229078613693999
    +	kno acc: 0.8512298959318827
    +	unk acc: 0.5853174603174603
    ::: Test scores (lemmas) :::
    +	all acc: 0.8271996615905245
    +	kno acc: 0.8962331201137171
    +	unk acc: 0.252465483234714
    ::: Test scores (pos) :::
    +	all acc: 0.8189509306260575
    +	kno acc: 0.8538261075574508
    +	unk acc: 0.5285996055226825
    ::: ended :::

Since I was not satisfied, I tried reloading and training again for 50 more epochs. I was then very surprised to see the various rates, which had not increased noticeably for some epochs, starting to increase 1 % per epoch, as well as training time going up. I wondered why, but now I think it is because of a «bug» in my loading function (?) which set learning rate very high (notice the change in learning rate and training time):

    -> epoch  150 ...
	    - Lowering learning rate > was: 1.0 , now: 0.5
    Epoch 1/1
    37853/37853 [==============================] - 266s - loss: 1.3604 - lemma_out_loss: 0.9304 - pos_out_loss: 0.4300   
    ::: Train scores (lemmas) :::
    +	all acc: 0.992365202229678
    +	kno acc: 0.992365202229678
    +	unk acc: 0.0
    ::: Dev scores (lemmas) :::
    +	all acc: 0.9234995773457312
    +	kno acc: 0.9739829706717124
    +	unk acc: 0.5
    ::: Train scores (pos) :::
    +	all acc: 0.983462341161863
    +	kno acc: 0.983462341161863
    +	unk acc: 0.0
    ::: Dev scores (pos) :::
    +	all acc: 0.9275147928994083
    +	kno acc: 0.9477294228949859
    +	unk acc: 0.7579365079365079
    ::: Test scores (lemmas) :::
    +	all acc: 0.9141285956006768
    +	kno acc: 0.9701492537313433
    +	unk acc: 0.4477317554240631
    ::: Test scores (pos) :::
    +	all acc: 0.9282994923857868
    +	kno acc: 0.9559346126510305
    +	unk acc: 0.6982248520710059

Have you encountered something like that in your trainings? How high is lr supposed to be at startup ?
In my case, it was at `0.009999999776482582` for the original run, and at `1` for the second (after loading). I guess it has something to do with the optimizer used ?

**Comment by mikekestemont** [comment]
Aha, this is because I didn't correctly anticipate the retraining scenario. Right now, the reloaded models don't explicitly get assigned a fresh learning, so they will switch to some sort of default. This means that we should pipe in the learning rate from the config file when reloading the model.

**Comment by Jean-Baptiste-Camps** [comment]
Adding lrate to config would be great, I think, also because it will be easier to manipulate it and benchmark different values. I'm tempted to try very high lrate now, seeing the results it might give in some cases. Note that when the model is saved, the value of learning rate should be the current one, I think, not the original config one, if we want to resume training from exactly where it was when it was stopped.

**Comment by emanjavacas** [comment]
This is problematic with adaptive optimizers, since there you not only need the learning rate but also the accumulated gradient per parameter..., so you would have to either start afresh with a lower learning rate, or serialize the optimizer.

**Comment by PonteIneptique** [comment]
We need to take this over now that lr is being set up.

**Comment by PonteIneptique** [comment]
@Jean-Baptiste-Camps what's your status with the current fix ?
